### ISSUE TITLE
netdata doesn't restart mysql connection when mysql-server is restarted

### ISSUE THREAD

**Issue by dseira** [firehol/netdata#1155]
Hi,

I don't know if it is a bug or is the normal behaviour. The problem comes when netdata is monitoring a mysql server; in that case, if I reboot the mysql-server the netdata not longer monitor the server:

![seleccion_022](https://cloud.githubusercontent.com/assets/378158/19638916/50ab34a4-99d6-11e6-8e62-0c2bc991e7d2.png)

As you can see in the previous screenshot the top graph is working but the graph below doesn't. It seems that the netdata doesn't reconnect to the mysql.

To fix this, I have to restart the netdata itself; and all work as expected:

![seleccion_023](https://cloud.githubusercontent.com/assets/378158/19638985/9e5398e0-99d6-11e6-8f04-ee3a1354bfcd.png)

Is this the normal behaviour or is a bug?

**Comment by ktsaou** [comment]
In `/etc/netdata/python.d/mysql.conf` there is a setting to control after how many retries to re-connect, netdata gives up:

```
# retries sets the number of retries to be made in case of failures.
# If unset, the default for python.d.plugin is used.
# Attempts to restore the service are made once every update_every
# and only if the module has collected values in the past.
# retries: 5
```

It is commented. The internal default is 60: https://github.com/firehol/netdata/blob/0f9f5e9bbf7365b13779b806111467171177402a/python.d/mysql.chart.py#L27

So, netdata will give up, if it cannot reconnect after 60 seconds (with per second updates, one retry is made every second).

If your mysql server takes too long to restart, I suggest to increase this accordingly.

**Comment by dseira** [comment]
I've modified the `/etc/netdata/python.d/mysql.conf` with:

> update_every: 1
retries: 120

But the result is the same; with a simple `service mysql restart` the netdata doesn't reconnect and the graph is stopped.

In the error.log appears this kind of messages:

> netdata: ERROR: child pid 476 killed by signal 15. (errno 9, Bad file descriptor)
netdata: ERROR: child pid 893 exited with code 1. (errno 9, Bad file descriptor)
netdata: ERROR: child pid 962 exited with code 1. (errno 9, Bad file descriptor)

I don't know if this can be related.

**Comment by ktsaou** [comment]
Is `python.d.plugin` running when this happens?

Can you search for the pids in error.log? Interesting is 476. I guess 893 and 962 are `alarm-notify.sh`, but what is 476?

**Comment by dseira** [comment]
I did more tests and the result is clear; I don't know why but `python.d.plugin` is killed and it doesn't start again.

it shows the message `Waiting a bit before starting it again.` but it is never restarted.

> 2016-10-25 20:00:43: netdata: ERROR: PLUGINSD: '/opt/netdata/usr/libexec/netdata/plugins.d/python.d.plugin' called DISABLE. Disabling it.
2016-10-25 20:00:43: netdata: INFO: PLUGINSD: '/opt/netdata/usr/libexec/netdata/plugins.d/python.d.plugin' on pid 32015 stopped after 7735 successful data collections (ENDs).
2016-10-25 20:00:43: netdata: ERROR: child pid 32015 killed by signal 15. (errno 9, Bad file descriptor)
2016-10-25 20:00:43: netdata: ERROR: PLUGINSD: '/opt/netdata/usr/libexec/netdata/plugins.d/python.d.plugin' exited with error code -1, but has given useful output in the past (7735 times). Waiting a bit before starting it again.
2016-10-25 20:00:51: alarm-notify.sh: FATAL: All notification methods are disabled. Not sending a notification.
2016-10-25 20:00:51: netdata: ERROR: child pid 32621 exited with code 1. (errno 9, Bad file descriptor)
2016-10-25 20:00:53: netdata: INFO: PLUGINSD: '/opt/netdata/usr/libexec/netdata/plugins.d/python.d.plugin' thread exiting

**Comment by ktsaou** [comment]
hm... the error message printed is wrong. I just pushed a commit to fix the error message, so that it will properly say that a plugin that called `DISABLE` will not be restarted.

Can you post of 2 minutes of log lines, before 20:00:43, i.e. lines after 19:58:43 ?

**Comment by dseira** [comment]
Sure. This is the log:

> 2016-10-25 19:57:02: netdata: ERROR: disabled cgroup cpu statistics.
2016-10-25 19:57:02: netdata: ERROR: Cannot read cgroups directory '/sys/fs/cgroup/blkio' (errno 2, No such file or directory)
2016-10-25 19:57:02: netdata: ERROR: disabled cgroup blkio statistics.
2016-10-25 19:57:02: netdata: ERROR: Cannot read cgroups directory '/sys/fs/cgroup/memory' (errno 2, No such file or directory)
2016-10-25 19:57:02: netdata: ERROR: disabled cgroup memory statistics.
2016-10-25 19:57:02: netdata: ERROR: Cannot read cgroups directory '/sys/fs/cgroup/devices' (errno 2, No such file or directory)
2016-10-25 19:57:02: netdata: ERROR: disabled cgroup devices statistics.
2016-10-25 19:57:02: netdata: ERROR: PROCFILE: Cannot open file '/proc/net/stat/nf_conntrack' (errno 2, No such file or directory)
2016-10-25 19:57:02: netdata: ERROR: PROCFILE: Cannot open file '/proc/net/ip_vs_stats' (errno 2, No such file or directory)
2016-10-25 19:57:02: netdata: ERROR: PROCFILE: Cannot open file '/proc/net/stat/synproxy' (errno 2, No such file or directory)
2016-10-25 19:57:02: netdata: INFO: Disabling /proc/net/rpc/nfsd threads histogram. It seems unused on this machine. It will be enabled automatically when found with data in it.
2016-10-25 19:57:02: netdata: INFO: Disabling /proc/net/rpc/nfsd read ahead histogram. It seems unused on this machine. It will be enabled automatically when found with data in it.
2016-10-25 19:57:02: netdata: INFO: Disabling /proc/net/rpc/nfsd v4 operations chart. It seems unused on this machine. It will be enabled automatically when found with data in it.
2016-10-25 19:57:11: alarm-notify.sh: INFO: not sending notification for CLEAR on 'mysql_local.queries.mysql_last_collected_secs' (last status was UNINITIALIZED)
2016-10-25 19:57:11: netdata: ERROR: child pid 32228 exited with code 1. (errno 9, Bad file descriptor)
2016-10-25 20:00:42: Prevented 11 log messages from displaying
2016-10-25 20:00:43: netdata: ERROR: PLUGINSD: '/opt/netdata/usr/libexec/netdata/plugins.d/python.d.plugin' called DISABLE. Disabling it.

**Comment by ktsaou** [comment]
ok, thanks!

@paulfantom check this one too. 1 second before python.d.plugin has logged this: https://github.com/firehol/netdata/blob/453e1d01fd32ce5ff1948e202669f9d71a827d47/python.d/python_modules/msg.py#L37

But apparently, these logs should have been printed. The log shown above does not include any other lines from python.d.plugin, so there was no reason to prevent these logs.

Also, this log line should state it comes from `python.d.plugin`.

Regarding the main issue, I think we cannot trace this issue, until the logging issue above is fixed.

@paulfantom could you please fix it?

**Comment by dseira** [comment]
Thanks, I will re-check the issue once the fix of the logging is done.

**Comment by ktsaou** [comment]
I think this must be resolved due to the recent changes in python.d.plugin.
Could you please check?

**Comment by dseira** [comment]
Right now netdata is showing the graphs and if the mysql server is rebooted the graphs continue working but in the logs I can see the following lines:

> 2016-11-21 11:20:22: python.d ERROR: mysql_local failed to collect data - 18 retries left - penalty: 0 sec
2016-11-21 11:20:23: python.d ERROR: mysql_local failed to collect data - 17 retries left - penalty: 0 sec
2016-11-21 11:20:24: python.d ERROR: mysql_local failed to collect data - 16 retries left - penalty: 0 sec
2016-11-21 11:20:25: python.d ERROR: mysql_local failed to collect data - 15 retries left - penalty: 0 sec
2016-11-21 11:20:26: python.d ERROR: mysql_local failed to collect data - 14 retries left - penalty: 0 sec
2016-11-21 11:20:27: python.d.plugin: throttling further log messages for 2373 seconds
2016-11-21 11:20:31: python.d ERROR: mysql_local failed to collect data - 50 retries left - penalty: 30.0 sec
2016-11-21 11:20:40: python.d ALERT: mysql_local failed to collect data for 60 times - increasing penalty to 30.0 sec and trying again

Even if netdata is showing the graphs as usual.

In the configuration file there is only one connection:

> tcp:
  name     : 'local'
  user     : 'root'
  pass     : 'pass1'
  host     : 'localhost'
  port     : '3306'

All other configurations are commented.

It seems that it is trying to reconnect but I don't know why because it is already connected (it is showing the graphs) and there are no other configuration options.

**Comment by ktsaou** [comment]
hm... could you please check that only one `python.d.plugin` is running? I believe a change made lately, prevents `python.d.plugin` from exiting when netdata restarts, but I couldn't reproduce it so far.

If I am right, you should have 2 `python.d.plugin` processes running, one under the netdata process tree and one alone.

**Comment by ktsaou** [comment]
yeap! I found a host with this situation, so I was able to reproduce it.

It is now fixed. Actually I had added code to handle all the cases where a module fails to collect data properly. These were fixed, and python.d.plugin is properly retrying to collect data.

But, I forgot the case of netdata exiting. In that case, python.d.plugin was staying alive. This problem was not visible on systemd systems, since systemd kills the whole process tree when the main process exits.

It is now fixed.
